command/agent_test.go

```diff
@@ -1691,3 +1695,304 @@ vault {
+				exitOnRetryFailure = fmt.Sprintf("exit_on_retry_failure = %t", *tc.exitOnRetryFailure)
+			}
+			templateConfig := fmt.Sprintf(`
+template_config = {
+	%s
+}
+`, exitOnRetryFailure)
+
+			template := fmt.Sprintf(`
+template {
+	contents = <<EOF
+%s
+EOF
+	destination = "%s/render_0.json"
+	error_on_missing_key = %t
+}
+`, tc.templateContents, tmpDir, tc.templateErrorOnMissingKey)
+
+			config := fmt.Sprintf(`
+# auto-auth stanza
+%s
+
+vault {
+	address = "%s"
+	tls_skip_verify = true
+	retry {
+		num_retries = 3
+	}
+}
+
+# listener stanza
+%s
+
+# template_config stanza
+%s
+
+# template stanza
+%s
+`, autoAuthConfig, serverClient.Address(), listenConfig, templateConfig, template)
+
+			configPath := makeTempFile(t, "config.hcl", config)
+			defer os.Remove(configPath)
+
+			// Start the agent
+			ui, cmd := testAgentCommand(t, logger)
+			cmd.startedCh = make(chan struct{})
+
+			// Channel to let verify() know to stop early if agent
+			// has exited
+			cmdRunDoneCh := make(chan struct{})
+			var exited bool
+
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			var code int
+			go func() {
+				code = cmd.Run([]string{"-config", configPath})
+				close(cmdRunDoneCh)
+				wg.Done()
+			}()
+
+			verify := func() error {
+				t.Helper()
+				// We need to poll for a bit to give Agent time to render the
+				// templates. Without this this, the test will attempt to read
+				// the temp dir before Agent has had time to render and will
+				// likely fail the test
+				tick := time.Tick(1 * time.Second)
+				timeout := time.After(15 * time.Second)
+				var err error
+				for {
+					select {
+					case <-cmdRunDoneCh:
+						exited = true
+						return nil
+					case <-timeout:
+						return fmt.Errorf("timed out waiting for templates to render, last error: %w", err)
+					case <-tick:
+					}
+					// Check for files rendered in the directory and break
+					// early for shutdown if we do have all the files
+					// rendered
+
+					//----------------------------------------------------
+					// Perform the tests
+					//----------------------------------------------------
+
+					if numFiles := testListFiles(t, tmpDir, ".json"); numFiles != 1 {
+						err = fmt.Errorf("expected 1 template, got (%d)", numFiles)
+						continue
+					}
+
+					fileName := filepath.Join(tmpDir, "render_0.json")
+					var c []byte
+					c, err = ioutil.ReadFile(fileName)
+					if err != nil {
+						continue
+					}
+					if strings.TrimSpace(string(c)) != tc.expectTemplateRender {
+						err = fmt.Errorf("expected='%s', got='%s'", tc.expectTemplateRender, strings.TrimSpace(string(c)))
+						continue
+					}
+					return nil
+				}
+			}
+
+			err = verify()
+			close(cmd.ShutdownCh)
+			wg.Wait()
+
+			switch {
+			case (code != 0 || err != nil) && tc.expectError:
+				if exited != tc.expectExit {
+					t.Fatalf("expected program exit to be '%t', got '%t'", tc.expectExit, exited)
+				}
+			case code == 0 && err == nil && !tc.expectError:
```

Comment:
Should we still test if we exited in this case?

Reply:
This is the happy path. The exit code was 0 and we didn't expect an error.

Reply:
I added an additional check within the happy path to make sure that we are not exiting early for any reason.